File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        
        int row_number = matrix.size();
        int col_number = matrix[0].size();
        int sRow = 0;
        int eRow = row_number - 1; 
        int sCol = 0;
        int eCol = col_number - 1;
        int total = row_number * col_number;
        int count = 0;
        vector<int> ans;
        
        while(sRow <= eRow && sCol <= eCol){
            
            //push first row
            for(int i = sCol; i<=eCol && count < total; i++){
                ans.push_back(matrix[sRow][i]);
                count++;
            }
            sRow++;
            
            //push last col
            for(int i = sRow; i <= eRow && count < total; i++){
                ans.push_back(matrix[i][eCol]);
                count++;
            }
            eCol--;
            
            //push last row
            for(int i = eCol; i >= sCol && count < total; i--){
                ans.push_back(matrix[eRow][i]);
                count++;
            }
            eRow--;
            
            //push first col
            for(int i = eRow; i>=sRow && count < total; i--){
                ans.push_back(matrix[i][sCol]);
                count++;
            }
            sCol++;
        }
        return ans;
    }
};

File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {

        List<Integer> ans = new ArrayList<>();

        int rows = matrix.length;
        int cols = matrix[0].length;
        int firstRow = 0;
        int lastRow = rows - 1;
        int firstCol = 0;
        int lastCol = cols - 1;

        while (firstRow <= lastRow && firstCol <= lastCol) {

            // first row
            for (int col = firstCol; col <= lastCol; col++) {
                ans.add(matrix[firstRow][col]);
            }
            firstRow++;

            // last column
            for (int row = firstRow; row <= lastRow; row++) {
                ans.add(matrix[row][lastCol]);
            }
            lastCol--;

            if (firstRow <= lastRow) {
                // last row
                for (int col = lastCol; col >= firstCol; col--) {
                    ans.add(matrix[lastRow][col]);
                }
                lastRow--;
            }

            if (firstCol <= lastCol) {
                // first column
                for (int row = lastRow; row >= firstRow; row--) {
                    ans.add(matrix[row][firstCol]);
                }
                firstCol++;
            }
        }

        return ans;
    }
}
